### ISSUE TITLE
lektor build and lektor deploy do not detect changes of macros and templates

### ISSUE THREAD

**Issue by torian257x** [lektor/lektor#410]
neither in backend, nor from command line.

I am using rsync to deploy

big show stopper

If I update a macro file or a page template file and do not visit it via browser, e.g. I do not run

`lektor server`

then go to the url

I do not get updates at all of the files

`yes | lektor clean && lektor build && lektor deploy`

doesn't work

`lektor build && lektor deploy`

doesn't work

`lektor build --output-path ./build && lektor deploy --output-path ./build`

doesn't work

what means 'doesnt work`?

doesn't work means, the new file version of the template file is not used, the old one is deployed instead.

what does work?

if I recompile my css files from the scss originals, that file change is recognized in the assets folder.

what does work 2?

if I use

`lektor server`

then visit the file that makes use of a changed template, it will update to the new version, and then running `lektor build && lektor deploy` will deploy that newly correctly created version of that file.

Is that a good workaround?
No! If I change a macro, do you want me to revisit _all_ blog posts _each_ time? this needs to be fixed

Do the admin backend cms "refresh build" work?

No! neither they work.

there is more discussion on gitter

**Comment by nixjdm** [comment]
Thanks for finding and posting this @Jossnaz. Certainly sounds bad. @smitty42, tagging you if you can get to this faster than I can.

**Comment by wgwz** [comment]
Can you post details about the version of Lektor you are using and how it was installed? I use roughly the same workflow, without any problems AFAIK.

**Comment by torian257x** [comment]
Lektor, version 3.0.1

i followed:

https://www.getlektor.com/docs/installation/

so

`sudo apt-get install python-dev libssl-dev libffi-dev`
`sudo apt-get install imagemagick`


actually thinking about it, because it is highlighted this way, I think I first ran:

`curl -sf https://www.getlektor.com/install.sh | sh`

and then afterwards added the missing packages when things failed and tried again.

what surprises me is that css changes are detected. But template changes aren't

**Comment by torian257x** [comment]
you cannot reproduce? It is lektor with python 2.7 btw on ubuntu 14.04

**Comment by torian257x** [comment]
when I run

`$ lektor build | grep index.html`

the pages I need updated are not getting listed...

**Comment by torian257x** [comment]
alright I can reproduce the bug like this:


I have a content 

`mypage1`

that uses a model

`mypage1-model`

`mypage1-model` content was this:


```
[model]
name = mypage1-model
label = Federation
hidden = yes


[pagination]
enabled = yes
per_page = 10
```

note, there is pagination.

then there is a template

`mypage1-model.html`

does _not_ get updated on build. On direct visit via server yes works.

to make it work I had to add this:

```
[model]
name = mypage1-model
label = Federation
hidden = yes


[pagination]
enabled = yes
per_page = 10

[children]
replaced_with = site.get('/').pagination.items.filter(F._model == 'blog-post')

```

note the children entry that got added, it is a random entry I had for children somehwere. Probably different values work too. Without this, it won't work.

The pagination exists generally for iterating, I think I had errors when traversing the records and trying to access pagination when it wasn't even added to the model resulting in a fatal error, so I just added it to all models.

I consider this a bug. The pagination iteration thing is another issue, but the records should get built with simple pagination added, no? or at the very least show an error.

**Comment by nixjdm** [comment]
@Jossnaz I'm curious, there were some improvements to the file watcher since 3.0.1 that are in master. Can you check if this bug exists in master too? You'll need to go though a installation with pip instead of the curl route to try it out.

**Comment by torian257x** [comment]
just tried, the bug still is in master

**Comment by jschwartzenberg** [comment]
I have the same problem with the pages below this page: http://www.terraflora.design/projects/
Anything directly below the root seems fine. It's just the deeper pages that aren't generated by a 'lektor build' only 'lektor server' and accessing the page through the server does the trick.

**Comment by ghost** [comment]
Hi everyone,

I figured that this is an old issue, however I think I just ran into the same problem:

I was lazy and only worked with the standard `page` model; since some of my pages do have children, I configured the `pagination` and `children` sections.
Pages that did not have children were not built by `lektor build` and only updated when visiting them via `lektor server`.

All pages were built when I removed the `pagination` (except of course when I used pagination in the templates). To fix it, I added a second model for everything that holds children.
@torian257x, @jschwartzenberg: not sure if this was exactly your problem or if you still experience that, can you confirm?

**Comment by dairiki** [comment]
What's going on in @Schubisu's report is this: [`Pagination.pages`][1] — the number of pages needed for the paginated page —  returns zero if a paginated page has no children.  The logic in [`PageBuildProgram`][2] then, dutifully, produces no pages for that childless paginated page.

[Edit: It may be [`PaginationConfig.count_pages`][3] which is causing trouble by returning zero pages.]

Upon first look, I think the correct solution is to change `Pagination.pages` so that it always returns at least one.  We always do want at least one page, even if there are no children, right?

(I haven't looked carefully enough to tell if this is the same issue as was reported earlier in this thread.)

[1]: <https://github.com/lektor/lektor/blob/6275719b6e3a224b2beb585b37cdb0e83b25992e/lektor/pagination.py#L28>
[2]: <https://github.com/lektor/lektor/blob/6275719b6e3a224b2beb585b37cdb0e83b25992e/lektor/build_programs.py#L191>
[3]: <https://github.com/lektor/lektor/blob/f273c6918b5f1043a6b09980de0a82f758033929/lektor/datamodel.py#L72>

I will try to produce a PR for this in the next couple of days.

**Comment by dairiki** [comment]
I think this was fixed in #952 (which is included in Lektor 3.3.0).

Closing.
